---
title: "A Pan Cancer Atlas"
output:
  pdf_document: default
  html_notebook: default
---
```{r}
library(dplyr)
library(ggplot2)
library(reshape2)
library(DESeq2)
library(DEGreport)
library(WGCNA)
library(rvest)

```

Using the high throughput data and employing transcriptomics analyses on the targeted dataset, we can evaluate the differentially expressed genes (DEGs), which can affect many biological pathways through their dysregulation. However, these studies cannot describe the quantitative association between genes and failed to give us a deep understanding of complex hierarchical networks and how the genes in critical pathways interact. Alternatively, we can utilize better methods to understand that quantitative association targets, not a set of candidate genes but a network of genes associated with biological processes. For instance, we can employ protein-protein or gene-gene interaction (PPI and GGI), annotated pathways (e.g., KEGG and PANTHER), and gene ontology terms. In this study, we used the RNA co-expression network analysis to find our datasets' most critical gene cluster. Weighted Gene Co-Expression Network Analysis (WGCNA). We aim to evaluate the DEGs and PPI analysis on TCGA datasets in twelve cancer types and study those candidate genes in the associated pathways. Therefore, by using WGCNA, we can score those findings based on the merged and unmerged gene clusters that WGCNA evaluated. Moreover, we can use a machine learning algorithm trained based on the adjacency matrix, which WGCNA constructed from, and associate that module with clinical data. This module will aptly explain at the end of this proposal. 
First and foremost, we should evaluate our DEGs. In this proposal, we use the LUAD-TCGA project, but the table of the candidate projects of twelve cancer types will be provided in the following.

```{r}
setwd("~/desktop/THCA")
ex <- read.delim("~/Desktop/THCA/ex_LUAD.tsv")
cn <- ex #backup

rownames(cn) <- cn$gene_id
cn <- cn[,-1:-2]

# Duplicated Counts from patients
cn <- cn[,-c(178,223,226,246,339,428,485,474)]
```


```{r}
#sample types
sampleID <- read.delim("~/desktop/THCA/TCGA-LUAD/Data/gdc_sample_sheet.2022-06-11.tsv")
i1 <- sampleID$Data.Type == "Gene Expression Quantification"
sampleID <- sampleID[i1, , drop = F]
sampleID <- sampleID[,7:8]
sampleID <- distinct(sampleID, Sample.ID, .keep_all = T)
a <- colnames(cn)
for(i in 1:3){a <- sub('\\.', '-', a)}
b <- data.frame(IDs = a)
c <- merge(b, sampleID, by.x = "IDs", by.y = "Sample.ID", sort = F)
cc <- data.frame(sample = sub("Primary Tumor", "PRT", c$Sample.Type))
cc <- data.frame(sample = sub("Solid Tissue Normal", "STN", cc$sample))
for(i in 1:3){c$IDs <- sub('\\-', '.', c$IDs)}
colnames(cn) <- cc$sample
# we reached to our tidy data set

```
```{r}
con <- factor(paste0(c$Sample.Type, seq_along(c$Sample.Type)))
gr <- factor(cc$sample)
#colSums(cn)
colData <- data.frame(condition = con,group = gr , type = "paired-end")
cds <- DESeqDataSetFromMatrix(cn, colData, design = ~group) 
cds <- DESeq(cds) #Constructing DESeq2  dataset
cnt <- log2(1+(counts(cds, normalize = T))) #getting normalized counts `count matrix`
```

